### PR TITLE
Hook GradleBuild.withConnection into Eclipse services

### PR DESCRIPTION
**Pull request by donat**

### CODE REVIEW

**Comment by donat** [comment]
@marcphilipp I have an open question on this PR. Instead of rethrowing exceptions, maybe we should introduce a `ConnectionResult<T>` interface where we return an `IStatus` along with the result (`Optional<T>`) for both successful and failed executions.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java [hunk 1]
```diff
@@ -62,14 +62,37 @@
...
+     *
+     * <p>
+     * This method instantiates a new connection to the Tooling API, pre-configures it with IDE
+     * services and executes the target action. Clients can use this method to load models and
+     * execute tasks, tests and custom build actions without explicitly configuring
+     * inputs/outputs/cancellation/etc.
+     *
      * <p>
-     * TODO (donat) explain how the action is executed (refer to Tooling API documentation)
+     * The following sections will show examples how can this API be used.
```
Comment (marcphilipp) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java]: :o: I'd remove the `will`

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java [hunk 2]
```diff
@@ -62,14 +62,37 @@
...
      * TODO (donat) explain common use-cases: task execution, test execution, and model loading
-     * TODO (donat) explain how the project connection is preconfigured: outputs and outputs to IDE console, progress and cancellation are wired to the monitor object
-     * TODO (donat) document exception handling
-     * TODO (donat) document scheduling rules
      * TODO (donat) document how to load custom models (here or link to external documentation)
      *
+     * <p>
+     * The following {@link ProjectConnection} methods that can't be wired into the Eclipse
```
Comment (marcphilipp) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java]: :x: remove the extra `that`

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java [hunk 3]
```diff
@@ -130,19 +132,28 @@ public GradleConnectionOperation(Function<ProjectConnection, ? extends T> action
...
-            ProjectConnection connection = connector.connect();
+            ProjectConnection connection = IdeAttachedProjectConnection.newInstance(tokenSource, getGradleArguments(), monitor);
             try {
                 this.result = this.action.apply(connection);
             } finally {
                 connection.close();
             }
         }
 
+        private  GradleArguments getGradleArguments() {
```
Comment (marcphilipp) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java]: nit: extra space after `private`

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/ConnectionAwareLauncherProxy.java [hunk 4]
```diff
@@ -44,7 +44,7 @@
  * @author Donat Csikos
  */
 @SuppressWarnings("unchecked")
-final class ConnectionAwareLauncherProxy implements InvocationHandler {
+public final class ConnectionAwareLauncherProxy implements InvocationHandler {
```
Comment (marcphilipp) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/ConnectionAwareLauncherProxy.java]: ❓ Why was this changed?

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/IdeAttachedProjectConnection.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    public <T> ModelBuilder<T> model(Class<T> modelType) {
+        return configuratOperation(() -> this.delegate.model(modelType));
+    }
+
+    @Override
+    public <T> BuildActionExecuter<T> action(BuildAction<T> buildAction) {
+        return configuratOperation(() -> this.delegate.action(buildAction));
+    }
+
+    private <T extends LongRunningOperation> T configuratOperation(Supplier<T> operationSupplier) {
```
Comment (marcphilipp) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/IdeAttachedProjectConnection.java]: :x: The method name has a typo. Moreover, why does it take a `Supplier` only to immediately call `get()` on it? Wouldn't it be easier to just pass a `T` instance?

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/IdeAttachedProjectConnection.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        throw new UnsupportedOperationException(explainUsage("getModel(Class)", "model(Class)"));
+    }
+
+    @Override
+    public <T> void getModel(Class<T> modelType, ResultHandler<? super T> handler) throws IllegalStateException {
+        throw new UnsupportedOperationException(explainUsage("getModel(Class, ResultHandler)", "model(Class)"));
+    }
+
+    private static String explainUsage(String methodSignature, String alternativeSignature) {
+        return "Cannot call ProjectConnection." + methodSignature + " as it is not possible to hook it's progress into the IDE. Use ProjectConnection." + alternativeSignature + " instead";
```
Comment (marcphilipp) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/IdeAttachedProjectConnection.java]: :x: `it's` -> `its`

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java [hunk 7]
```diff
@@ -130,19 +132,28 @@ public GradleConnectionOperation(Function<ProjectConnection, ? extends T> action
...
+            org.eclipse.buildship.core.internal.configuration.BuildConfiguration config = DefaultGradleBuild.this.gradleBuild.getBuildConfig();
+            return GradleArguments.from(
```
Comment (marcphilipp) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java]: ❓ Why doesn't this call `config.toGradleArguments()`?

File: org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/GradleBuildConnectionConcurrencyTest.groovy [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        setup:
+        // import a sample project
+        File location = dir('GradleBuildConnectionConcurrencyTest')
+        BuildConfiguration buildConfiguration = BuildConfiguration.forRootProjectDirectory(location).build()
+        GradleBuild gradleBuild = GradleCore.workspace.createBuild(buildConfiguration)
+        Job modelQuertJob = new ModelQueryJob(gradleBuild)
+
+        when:
+        IWorkspaceRunnable modelQueryOperation = {
+            modelQuertJob.schedule()
```
Comment (marcphilipp) [inline org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/GradleBuildConnectionConcurrencyTest.groovy]: :x: `modelQueryJob`

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java [hunk 9]
```diff
@@ -62,14 +62,37 @@
...
+     * services. For that reason the target action cannot use them, otherwise an
+     * {@link UnsupportedOperationException} is thrown.
+     * <ul>
+     * <li>{@link ProjectConnection#model(Class)}</li>
+     * <li>{@link ProjectConnection#getModel(Class, org.gradle.tooling.ResultHandler))}</li>
+     * <li>{@link ProjectConnection#action()}</li>
```
Comment (marcphilipp) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java]: ❓ Should these be deprecated because they have alternatives that are safe to use?

**Comment by donat** [comment]
Makes sense @marcphilipp, I just wanted to point out the possible alternative.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/ConnectionAwareLauncherProxy.java (hunk 4 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/ConnectionAwareLauncherProxy.java]: It's just a residue from my fiddling with the test coverage. It should not be there.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java (hunk 9 above)
Comment (oehme) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java]: ❌ I see no reason why these methods wouldn't work. We should support all of them.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java (hunk 1 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java]: Fixed

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java (hunk 2 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java]: Fixed

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java (hunk 3 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java]: Fixed

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/ConnectionAwareLauncherProxy.java (hunk 4 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/ConnectionAwareLauncherProxy.java]: Fixed

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/IdeAttachedProjectConnection.java (hunk 5 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/IdeAttachedProjectConnection.java]: Fixed

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/IdeAttachedProjectConnection.java (hunk 6 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/IdeAttachedProjectConnection.java]: Fixed

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java (hunk 7 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java]: Fixed

File: org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/GradleBuildConnectionConcurrencyTest.groovy (hunk 8 above)
Comment (donat) [reply org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/GradleBuildConnectionConcurrencyTest.groovy]: Fixed

**Comment by donat** [comment]
@oehme I've added support for all `ProjectConnection` methods.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java (hunk 9 above)
Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/GradleBuild.java]: Fixed

**Comment by oehme** [comment]
Nice work!
